### PR TITLE
Codebase Update via Noah's Refactored DDPM Notebook

### PR DESCRIPTION
**Pull request by mateibejan1**
The following modules have been updated:

1. Conditioned DDPM (with extra p2_weighting function for loss optimization, still need to figure out how to apply it to our use case).
2. Diffusion base class.
2. Metrics.

### CODE REVIEW

File: src/models/networks/unet_bitdiffusion.py [hunk 1]
```diff
@@ -1,39 +1,42 @@
...
+
 class UNet(nn.Module):
     def __init__(
         self,
         dim,
-        init_dim = None,
+        init_dim=None,
         dim_mults=(1, 2, 4, 8),
-        channels = 3,
-        bits = BITS,
```
Comment (SauravMaheshkar) [inline src/models/networks/unet_bitdiffusion.py:10]: What is `BITS` ? I'm assuming it's meant to be a type hint but it isn't defined anywhere in this script.

File: src/models/networks/unet_bitdiffusion.py [hunk 2]
```diff
@@ -1,39 +1,42 @@
...
-        self.channels = channels *2
+        channels *= bits  # lucas
+        self.channels = channels * 2
 
         input_channels = channels * 2
-        #input_channels =16
+        # input_channels =16
 
-        
         init_dim = default(init_dim, dim)
```
Comment (SauravMaheshkar) [inline src/models/networks/unet_bitdiffusion.py:27]: What is `default` ?

File: src/models/diffusion/ddim.py [hunk 3]
```diff
@@ -45,38 +46,40 @@ def device(self):
...
+        times = torch.stack((times[:, :-1], times[:, 1:]), dim=0)
+        times = times.unbind(dim=-1)
         return times
 
     @torch.no_grad()
-    def ddim_sample(self, shape, classes, time_difference = None):
+    def ddim_sample(self, shape, classes, time_difference=None):
         batch, device = shape[0], self.device
 
         time_difference = default(time_difference, self.time_difference)
```
Comment (SauravMaheshkar) [inline src/models/diffusion/ddim.py:59]: Again, `default` ?
